triggers/daml/Daml/Trigger/LowLevel.daml

```diff
@@ -244,3 +255,111 @@ data Commands = Commands
   { commandId : CommandId
   , commands : [Command]
   }
+
+-- Same as in Script; TODO factor
+data Free f a
+  = Pure a
+  | Free (f (Free f a))
+
+instance Functor f => Functor (Free f) where
+  fmap f (Pure a) = Pure (f a)
+  fmap f (Free x) = Free (fmap f <$> x)
+
+instance Functor f => Applicative (Free f) where
+  pure = Pure
+  Pure f <*> Pure a = Pure (f a)
+  Pure f <*> Free x = Free (fmap f <$> x)
+  Free x <*> my     = Free ((<*> my) <$> x)
+
+instance Functor f => Action (Free f) where
+  Pure a >>= f = f a
+  Free x >>= f = Free ((>>= f) <$> x)
+
+liftFC : Functor f => f a -> Free f a
+liftFC fa = Free (Pure <$> fa)
+
+lowerFC : Action f => Free f a -> f a
+lowerFC (Pure a) = pure a
+lowerFC (Free x) = x >>= lowerFC
+
+newtype StateT s m a = StateT { runStateT : s -> m (a, s) }
```

Comment:
Note that newtypes are not free in DAML so you are sometimes better off inlining this stuff (sadly). Not too worried about this for triggers for now but worth keeping in mind.

Reply:
Also I just realized that we used a newtype before as well … Maybe at least worth merging it to one layer of newtypes at some point instead of having `TriggerA` on top of `TriggerRule` on top of `StateT` …. Doesn’t have to be in this PR though.

Reply:
I use them for three things in this PR:

1. matching constraint signatures nicely,
2. letting the functor layer go up and down when I haven't settled on which layer will operate where. e.g. TriggerSetup very nearly got elided for the final design, but it was needed for the low-level initialState as you wanted, so it survived causing very little trouble at the DAML layer (at the Scala layer is another story), and
3. _not_ letting operations pollute upwards. For example it would be very undesirable for `ActionState TriggerAState TriggerA` to exist; it might even be desirable to remove `ActionTrigger TriggerA` so as to hide `submitCommands` from the high-level API.

(2) is naturally just scaffolding, but we should take care about (3). And of course damlc will just yell at you for not thinking about (1).

Reply:
Makes sense :+1: